plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/IsNullPushdownDataTypeTest.java

```diff
@@ -0,0 +1,211 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.clickhouse;
+
+import io.trino.Session;
+import io.trino.spi.type.Type;
+import io.trino.sql.planner.plan.FilterNode;
+import io.trino.sql.query.QueryAssertions;
+import io.trino.testing.MaterializedResult;
+import io.trino.testing.QueryRunner;
+import io.trino.testing.datatype.ColumnSetup;
+import io.trino.testing.datatype.DataSetup;
+import io.trino.testing.sql.TemporaryRelation;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.IntStream;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static io.trino.plugin.clickhouse.IsNullPushdownDataTypeTest.PushdownImplementation.ALWAYS;
+import static io.trino.plugin.clickhouse.IsNullPushdownDataTypeTest.PushdownImplementation.CONNECTOR_EXPRESSION_ONLY;
+import static io.trino.plugin.clickhouse.IsNullPushdownDataTypeTest.PushdownImplementation.TUPLE_DOMAIN_ONLY;
+import static io.trino.plugin.clickhouse.IsNullPushdownDataTypeTest.Scenario.IS_NOT_NULL;
+import static io.trino.plugin.clickhouse.IsNullPushdownDataTypeTest.Scenario.IS_NULL;
+import static java.lang.String.format;
+import static java.util.Collections.unmodifiableList;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.joining;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public final class IsNullPushdownDataTypeTest
+{
+    private static final int SPECIAL_COLUMNS = 1;
+
+    private final List<TestCase> testCases = new ArrayList<>();
+    private final PushdownImplementation pushdownImplementation;
+
+    private IsNullPushdownDataTypeTest(PushdownImplementation pushdownImplementation)
+    {
+        this.pushdownImplementation = pushdownImplementation;
+    }
+
+    public static IsNullPushdownDataTypeTest create(PushdownImplementation pushdownImplementation)
+    {
+        return new IsNullPushdownDataTypeTest(pushdownImplementation);
+    }
+
+    public IsNullPushdownDataTypeTest addRoundTrip(String inputType, String inputLiteral, Type expectedType, String expectedLiteral)
+    {
+        testCases.add(new TestCase(Optional.of(inputType), inputLiteral, Optional.of(expectedType), expectedLiteral));
+        return this;
+    }
+
+    public IsNullPushdownDataTypeTest execute(QueryRunner queryRunner, DataSetup dataSetup)
+    {
+        return execute(queryRunner, queryRunner.getDefaultSession(), dataSetup);
+    }
+
+    public IsNullPushdownDataTypeTest execute(QueryRunner queryRunner, Session session, DataSetup dataSetup)
+    {
+        checkState(!testCases.isEmpty(), "No test cases");
+        for (int specialColumn = 0; specialColumn < SPECIAL_COLUMNS; specialColumn++) {
+            checkArgument(!"NULL".equalsIgnoreCase(testCases.get(specialColumn).inputLiteral()));
```

Comment:
After going through the code, I got the intention of why we needed `SPECIAL_COLUMNS`. May be `NON_NULL_COLUMNS` makes more sense here. WDYT?

Reply:
I had `TestClickHouseConnectorTest#testTextualPredicatePushdown` in mind, where SPECIAL_COLUMNS are
```
                        unsupported_1 Point,
                        unsupported_2 Point,
                        some_column String,
```

Reply:
What if we could handle this special columns as part of this framework ?

Reply:
I think it's possible however I'm not sure how to make it look better that it is right now, because the type of the special columns may differ for different cases:
```
                .addTestCase("String", "'z'", VARCHAR, "CAST('z' AS varchar)") // special, non null column
                .addTestCase("String", "'z'", VARBINARY, "CAST('z' AS varbinary)") // special, non null column
```